From f98dd69d615875906d25b213da3a228e424f3df7 Mon Sep 17 00:00:00 2001
From: Sandy Huang <hjc@rock-chips.com>
Date: Wed, 23 Aug 2017 16:38:32 +0800
Subject: [PATCH] drm/panel: simple: add power-invert for some special hardwre

if power-invert exist the panel power need to disable ldo when
power on and enable ldo when power off otherwise it's opposite.

example:
	panel {
		...

		power-invert;

		...
	};

Change-Id: Ida5718d01044873cdd7c753c4e8b872dc1e52099
Signed-off-by: Sandy Huang <hjc@rock-chips.com>
---
 .../bindings/display/panel/simple-panel.txt   |  5 ++
 drivers/gpu/drm/panel/panel-simple.c          | 52 +++++++++++++++++--
 2 files changed, 52 insertions(+), 5 deletions(-)

diff --git a/Documentation/devicetree/bindings/display/panel/simple-panel.txt b/Documentation/devicetree/bindings/display/panel/simple-panel.txt
index ed6bfe3e459e..0e2c86bc6871 100644
--- a/Documentation/devicetree/bindings/display/panel/simple-panel.txt
+++ b/Documentation/devicetree/bindings/display/panel/simple-panel.txt
@@ -39,6 +39,11 @@ Optional properties when compatible is a dsi devices:
 	byte 2: packet payload length
 	byte 3 and beyond: number byte of payload
 
+Optional for some special hardwre:
+- power-invert:
+	if power-invert exist the panel power need to disable ldo when power on
+	and enable ldo when power off otherwise it's opposite.
+
 Example:
 
 	panel: panel {
diff --git a/drivers/gpu/drm/panel/panel-simple.c b/drivers/gpu/drm/panel/panel-simple.c
index ec7853c90838..c44789d23f79 100644
--- a/drivers/gpu/drm/panel/panel-simple.c
+++ b/drivers/gpu/drm/panel/panel-simple.c
@@ -101,6 +101,7 @@ struct panel_simple {
 	struct mipi_dsi_device *dsi;
 	bool prepared;
 	bool enabled;
+	bool power_invert;
 
 	struct device *dev;
 	const struct panel_desc *desc;
@@ -335,12 +336,32 @@ static int panel_simple_of_get_native_mode(struct panel_simple *panel)
 	return 1;
 }
 
-static int panel_simple_loader_protect(struct drm_panel *panel, bool on)
+static int panel_simple_regulator_enable(struct drm_panel *panel)
 {
 	struct panel_simple *p = to_panel_simple(panel);
-	int err;
+	int err = 0;
 
-	if (on) {
+	if (p->power_invert) {
+		if (regulator_is_enabled(p->supply) > 0)
+			regulator_disable(p->supply);
+	} else {
+		err = regulator_enable(p->supply);
+		if (err < 0) {
+			dev_err(panel->dev, "failed to enable supply: %d\n",
+				err);
+			return err;
+		}
+	}
+
+	return err;
+}
+
+static int panel_simple_regulator_disable(struct drm_panel *panel)
+{
+	struct panel_simple *p = to_panel_simple(panel);
+	int err = 0;
+
+	if (p->power_invert) {
 		err = regulator_enable(p->supply);
 		if (err < 0) {
 			dev_err(panel->dev, "failed to enable supply: %d\n",
@@ -351,6 +372,24 @@ static int panel_simple_loader_protect(struct drm_panel *panel, bool on)
 		regulator_disable(p->supply);
 	}
 
+	return err;
+}
+
+static int panel_simple_loader_protect(struct drm_panel *panel, bool on)
+{
+	int err;
+
+	if (on) {
+		err = panel_simple_regulator_enable(panel);
+		if (err < 0) {
+			dev_err(panel->dev, "failed to enable supply: %d\n",
+				err);
+			return err;
+		}
+	} else {
+		panel_simple_regulator_disable(panel);
+	}
+
 	return 0;
 }
 
@@ -394,7 +433,7 @@ static int panel_simple_unprepare(struct drm_panel *panel)
 	if (p->enable_gpio)
 		gpiod_direction_output(p->enable_gpio, 0);
 
-	regulator_disable(p->supply);
+	panel_simple_regulator_disable(panel);
 
 	if (p->desc && p->desc->delay.unprepare)
 		msleep(p->desc->delay.unprepare);
@@ -412,7 +451,7 @@ static int panel_simple_prepare(struct drm_panel *panel)
 	if (p->prepared)
 		return 0;
 
-	err = regulator_enable(p->supply);
+	err = panel_simple_regulator_enable(panel);
 	if (err < 0) {
 		dev_err(panel->dev, "failed to enable supply: %d\n", err);
 		return err;
@@ -580,6 +619,9 @@ static int panel_simple_probe(struct device *dev, const struct panel_desc *desc)
 		return err;
 	}
 
+	panel->power_invert =
+			of_property_read_bool(dev->of_node, "power-invert");
+
 	backlight = of_parse_phandle(dev->of_node, "backlight", 0);
 	if (backlight) {
 		panel->backlight = of_find_backlight_by_node(backlight);
-- 
2.35.3

